File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/dbn_simple_tutorial.m
% Simple DBN demonstration using one RBM layer.
% This example uses a toy dataset and a basic implementation of Contrastive Divergence.
close all; clear all; clc; 

% Parameters
numEpochs = 100;
learningRate = 0.05;
numHidden = 10;

% Generate toy data: 200 samples with 20 features (normalized)
X = rand(200, 20);

% Train RBM as the first layer of the DBN
[W, b_hidden, b_visible] = trainRBM(X, numHidden, numEpochs, learningRate);

% Obtain hidden representations (activations)
H = sigmoid(bsxfun(@plus, X * W, b_hidden));

% Visualize hidden activations using a heatmap
figure;
imagesc(H(1:50, :)); % visualize first 50 samples
colorbar;
title('Hidden Activations from the RBM Layer');
xlabel('Hidden Units');
ylabel('Sample Index');

% Display learned weight matrix as an image
figure;
imagesc(W);
colorbar;
title('Learned Weight Matrix');
xlabel('Hidden Units');
ylabel('Input Features');

function [W, b_hidden, b_visible] = trainRBM(X, numHidden, numEpochs, learningRate)
    % trainRBM trains a Restricted Boltzmann Machine using Contrastive Divergence.
    % Inputs:
    %   X           - Data matrix (n_samples x n_features)
    %   numHidden   - Number of hidden units
    %   numEpochs   - Number of training epochs
    %   learningRate- Learning rate for parameter updates
    % Outputs:
    %   W           - Weight matrix (n_features x numHidden)
    %   b_hidden    - Bias vector for hidden units (1 x numHidden)
    %   b_visible   - Bias vector for visible units (1 x n_features)
    
    [n_samples, n_features] = size(X);
    
    % Initialize parameters
    W = 0.1 * randn(n_features, numHidden);
    b_visible = zeros(1, n_features);
    b_hidden = zeros(1, numHidden);
    
    % Number of Contrastive Divergence steps
    k = 1;
    
    for epoch = 1:numEpochs
        % For each sample (for simplicity, process one sample at a time)
        for i = 1:n_samples
            v0 = X(i, :); % initial visible state
            
            % Compute probabilities for hidden units
            h0_prob = sigmoid(v0 * W + b_hidden);
            h0 = double(h0_prob > rand(1, numHidden));
            
            % Perform k steps of Gibbs sampling
            vk = v0;
            hk = h0;
            for step = 1:k
                % Sample visible units given hidden units
                vk_prob = sigmoid(hk * W' + b_visible);
                vk = double(vk_prob > rand(1, n_features));
                % Sample hidden units given visible units
                hk_prob = sigmoid(vk * W + b_hidden);
                hk = double(hk_prob > rand(1, numHidden));
            end
            
            % Compute gradients
            dW = (v0' * h0_prob) - (vk' * hk_prob);
            db_visible = v0 - vk;
            db_hidden = h0_prob - hk_prob;
            
            % Update parameters
            W = W + learningRate * dW;
            b_visible = b_visible + learningRate * db_visible;
            b_hidden = b_hidden + learningRate * db_hidden;
        end
        if mod(epoch, 10) == 0
            fprintf('Epoch %d complete.\n', epoch);
        end
    end
end

function y = sigmoid(x)
    y = 1 ./ (1 + exp(-x));
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/R/dbn_simple_tutorial.R
# dbn_simple_tutorial.R
# This script demonstrates a simple DBN using one RBM layer on toy data.
# It trains an RBM with Contrastive Divergence (CD-1) and visualizes the hidden activations and weight matrix.

rm(list=ls())
graphics.off()
set.seed(1)

# Parameters
numEpochs <- 100
learningRate <- 0.05
numHidden <- 10

# Generate toy data: 200 samples with 20 features (normalized)
X <- matrix(runif(200 * 20), nrow = 200, ncol = 20)

# Sigmoid activation function
sigmoid <- function(x) { 1 / (1 + exp(-x)) }

# Train RBM using CD-1
trainRBM <- function(X, numHidden, numEpochs, learningRate) {
  n_samples <- nrow(X)
  n_features <- ncol(X)
  W <- 0.1 * matrix(rnorm(n_features * numHidden), nrow = n_features)
  b_visible <- rep(0, n_features)
  b_hidden <- rep(0, numHidden)
  k <- 1  # number of Gibbs steps (CD-1)
  
  for (epoch in 1:numEpochs) {
    for (i in 1:n_samples) {
      v0 <- X[i, ]
      h0_prob <- sigmoid(v0 %*% W + b_hidden)
      h0 <- as.numeric(h0_prob > runif(numHidden))
      vk <- v0
      hk <- h0
      for (step in 1:k) {
        vk_prob <- sigmoid(hk %*% t(W) + b_visible)
        vk <- as.numeric(vk_prob > runif(n_features))
        hk_prob <- sigmoid(vk %*% W + b_hidden)
        hk <- as.numeric(hk_prob > runif(numHidden))
      }
      dW <- (outer(v0, h0_prob)) - (outer(vk, hk_prob))
      db_visible <- v0 - vk
      db_hidden <- h0_prob - hk_prob
      W <- W + learningRate * dW
      b_visible <- b_visible + learningRate * db_visible
      b_hidden <- b_hidden + learningRate * db_hidden
    }
    if (epoch %% 10 == 0) {
      cat(sprintf("Epoch %d complete.\n", epoch))
    }
  }
  list(W = W, b_hidden = b_hidden, b_visible = b_visible)
}

rbm <- trainRBM(X, numHidden, numEpochs, learningRate)

# Obtain hidden representations (activations)
H <- sigmoid(X %*% rbm$W + matrix(rep(rbm$b_hidden, nrow(X)), nrow = nrow(X), byrow = TRUE))

# Visualize hidden activations using a heatmap (first 50 samples)
heatmap(H[1:50, ], Rowv = NA, Colv = NA, scale = "none",
        main = "Hidden Activations from the RBM Layer")

# Display the learned weight matrix as an image
image(rbm$W, col = gray.colors(256), main = "Learned Weight Matrix",
      xlab = "Hidden Units", ylab = "Input Features")
